File: windows/Core/Km/KmLinkMonitor.cs
namespace BarelyReal.Core.Km;

/// Tracks "are KM frames flowing" on the receive side. Mirrors mac KmLinkMonitor.swift.
public sealed class KmLinkMonitor : IDisposable
{
    public sealed class Configuration
    {
        public TimeSpan LinkLostAfter { get; set; } = TimeSpan.FromMilliseconds(1500);
        public TimeSpan LockScreenAfter { get; set; } = TimeSpan.FromMilliseconds(5000);
        public TimeSpan PollInterval { get; set; } = TimeSpan.FromMilliseconds(100);
    }

    public enum Event
    {
        LinkLost,
        LinkRecovered,
        ShouldLockScreen
    }

    public event Action<Event, TimeSpan>? OnEvent;

    public bool IsLinkUp { get; private set; }
    public bool IsActive { get; private set; }

    private readonly Configuration _configuration;
    private readonly object _gate = new();
    private DateTime? _lastFrameAt;
    private bool _lockTriggered;
    private System.Threading.Timer? _timer;

    public KmLinkMonitor(Configuration? configuration = null)
    {
        _configuration = configuration ?? new Configuration();
    }

    public void Start()
    {
        Stop();
        IsActive = true;
        _timer = new System.Threading.Timer(_ => Tick(), null, _configuration.PollInterval, _configuration.PollInterval);
    }

    public void Stop()
    {
        _timer?.Dispose();
        _timer = null;
        IsActive = false;
        IsLinkUp = false;
        lock (_gate)
        {
            _lastFrameAt = null;
            _lockTriggered = false;
        }
    }

    public void NoteFrame()
    {
        bool recovered = false;
        lock (_gate)
        {
            _lastFrameAt = DateTime.UtcNow;
            _lockTriggered = false;
            if (!IsLinkUp)
            {
                IsLinkUp = true;
                recovered = true;
            }
        }
        if (recovered)
            OnEvent?.Invoke(Event.LinkRecovered, TimeSpan.Zero);
    }

    public void Dispose() => Stop();

    private void Tick()
    {
        if (!IsActive) return;

        DateTime? lastFrameAt;
        bool lockTriggered;
        lock (_gate)
        {
            lastFrameAt = _lastFrameAt;
            lockTriggered = _lockTriggered;
        }

        if (lastFrameAt is null) return;

        var silent = DateTime.UtcNow - lastFrameAt.Value;

        if (IsLinkUp && silent >= _configuration.LinkLostAfter)
        {
            IsLinkUp = false;
            OnEvent?.Invoke(Event.LinkLost, silent);
        }

        if (!lockTriggered && silent >= _configuration.LockScreenAfter)
        {
            lock (_gate) { _lockTriggered = true; }
            OnEvent?.Invoke(Event.ShouldLockScreen, silent);
        }
    }
}

File: mac/Core/KM/KmLinkMonitor.swift
import Foundation

/// Tracks "are KM frames flowing" on the receive side.
///
/// Wires:
///  - call `noteFrame()` on every successfully decoded inbound frame
///  - subscribe to `onLinkLost` / `onLinkRecovered`
///
/// Default thresholds match the BRP spec: 1500 ms = link lost, 5000 ms = optionally lock screen.
public final class KmLinkMonitor {
    public struct Configuration {
        public var linkLostAfter: TimeInterval
        public var lockScreenAfter: TimeInterval
        public var pollInterval: TimeInterval

        public init(linkLostAfter: TimeInterval = 1.5, lockScreenAfter: TimeInterval = 5.0, pollInterval: TimeInterval = 0.1) {
            self.linkLostAfter = linkLostAfter
            self.lockScreenAfter = lockScreenAfter
            self.pollInterval = pollInterval
        }
    }

    public enum Event: Equatable {
        case linkLost(silentFor: TimeInterval)
        case linkRecovered
        case shouldLockScreen
    }

    public var onEvent: ((Event) -> Void)?

    public private(set) var isActive = false
    public private(set) var isLinkUp = false

    private let configuration: Configuration
    private let queue: DispatchQueue
    private var timer: DispatchSourceTimer?
    private var lastFrameAt: Date?
    private var lockTriggered = false

    public init(configuration: Configuration = Configuration(),
                queue: DispatchQueue = DispatchQueue(label: "com.barelyreal.km-link-monitor")) {
        self.configuration = configuration
        self.queue = queue
    }

    public func start() {
        stop()
        isActive = true
        let timer = DispatchSource.makeTimerSource(queue: queue)
        timer.schedule(deadline: .now() + configuration.pollInterval,
                       repeating: configuration.pollInterval)
        timer.setEventHandler { [weak self] in self?.tick() }
        timer.resume()
        self.timer = timer
    }

    public func stop() {
        timer?.cancel()
        timer = nil
        isActive = false
        isLinkUp = false
        lastFrameAt = nil
        lockTriggered = false
    }

    public func noteFrame() {
        let now = Date()
        queue.async { [weak self] in
            guard let self else { return }
            self.lastFrameAt = now
            self.lockTriggered = false
            if !self.isLinkUp {
                self.isLinkUp = true
                self.onEvent?(.linkRecovered)
            }
        }
    }

    private func tick() {
        guard isActive, let lastFrameAt else { return }
        let silent = Date().timeIntervalSince(lastFrameAt)

        if isLinkUp, silent >= configuration.linkLostAfter {
            isLinkUp = false
            onEvent?(.linkLost(silentFor: silent))
        }

        if !lockTriggered, silent >= configuration.lockScreenAfter {
            lockTriggered = true
            onEvent?(.shouldLockScreen)
        }
    }
}
